### PR TITLE
fix report in a case of afterEach conditioanl fail

### LINKED ISSUE #1635: afterEach halts suite
This is a common pattern with tools like sinon or nock - to have afterEach do some standard verification at the end of each test.

I'm fairly sure this used to work - not sure what's changed.

``` js
var assert = require('assert');

describe("afterEach", function() {
    var verificationFlag;
    beforeEach(function() {
        verificationFlag = 0;
    });
    afterEach(function() {
        assert.equal(verificationFlag, 1);
    });

    it("0", function() { })
    it("1", function() { verificationFlag = 1 })
    it("2", function() { verificationFlag = 2 })
});
```

Expected output: 3 tests, 3 failures
Actual output: 1 test, 1 pass, 1 fail (from hook)

### LINKED ISSUE #3345: afterEach conditional error hook breaks stats and error reporting
### Prerequisites

- [x] Checked that your issue hasn't already been filed by cross-referencing [issues with the `faq` label](https://github.com/mochajs/mocha/issues?utf8=%E2%9C%93&q=is%3Aissue%20label%3Afaq%20)
- [x] Checked next-gen ES issues and syntax problems by using the same environment and/or transpiler configuration without Mocha to ensure it isn't just a feature that actually isn't supported in the environment in question or a bug in your code.
- [x] 'Smoke tested' the code to be tested by running it outside the real test suite to get a better sense of whether the problem is in the code under test, your usage of Mocha, or Mocha itself
- [x] Ensured that there is no discrepancy between the locally and globally installed versions of Mocha. You can find them with: `node node_modules/.bin/mocha --version`(Local) and `mocha --version`(Global). We recommend avoiding the use of globally installed Mocha.

### Description

Trying to fail a test with this.test.error() causes the test to be reported twice, as failed and as success

### Steps to Reproduce

Run the test at:

https://github.com/mochajs/mocha/wiki/HOW-TO:-Conditionally-fail-a-test-after-completion

**Expected behavior:**
```
  something
    ✓ should one
    1) should two

  1 passing (16ms)
  1 failing
```

**Actual behavior:**
```
  something
    ✓ should one
    ✓ should two
    1) should two

  2 passing (11ms)
  1 failing

```

### PR DESCRIPTION
**Pull request by outsideris**
### Description of the Change
In our [wiki](https://github.com/mochajs/mocha/wiki/HOW-TO:-Conditionally-fail-a-test-after-completion), describe  "Conditionally fail a test after completion".
But it behaves like the explanation. 
If a test failed after it already passed, the final report says both the test passed and the test failed. I referred #1944 .

### Alternate Designs
I don't have an idea now.

### Why should this be in core?
Users could get a correct report.

### Benefits
We can report correctly. It is confused if same test place on both pass and failure.

### Possible Drawbacks
It seems a broken change because report order could be changed in some edge case.
And this PR is inspired by #1944 , the PR's owner didn't sign CLA. I'm not sure whether we can make this code change or we should get permission from the one.

### Applicable issues
close #3345
close #1635

### CODE REVIEW

File: test/integration/hooks.spec.js [hunk 1]
```diff
@@ -49,4 +50,33 @@ describe('hooks', function() {
       done();
     });
   });
+
+  it('can fail a test in an afterEach hook', function(done) {
```
Comment (boneskull) [inline test/integration/hooks.spec.js]: wouldn't it be easier to just look at the output of `runMochaJSON()`?

File: test/integration/hook-err.spec.js [hunk 2]
```diff
@@ -39,7 +39,7 @@ describe('hook error handling', function() {
   describe('after each hook error', function() {
     before(run('hooks/afterEach-hook-error.fixture.js'));
     it('should verify results', function() {
-      assert.deepEqual(lines, ['test 1', 'after', bang + 'test 3']);
+      assert.deepEqual(lines, ['test 1', 'after', bang, 'test 3']);
```
Comment (boneskull) [inline test/integration/hook-err.spec.js]: is this related?

Comment (outsideris) [reply test/integration/hook-err.spec.js]: https://github.com/mochajs/mocha/blob/8368722830ed80e03a52d29155310cdc7b6fc123/test/integration/hook-err.spec.js#L159 
In here,`splitRegExp` is `[\n.]+`.
So, prev result is `.!test 3`, but now the result is `!.test 3`.
Therefore, `!` and `test 3` is separated.

File: test/integration/hooks.spec.js (hunk 1 above)
Comment (outsideris) [reply test/integration/hooks.spec.js]: I changed to use it.

**Comment by outsideris** [comment]
How about `semver-minor`?
I think it is correct behavior, but it works as previous behavior for a long time even if our wiki describes it.
Now, checking test passed ran in `afterEach`, not `run`.
This change pass count/failure count.

**Comment by plroebuck** [comment]
Can we fix spelling in PR title?

**Comment by outsideris** [comment]
I fixed type in commit message as well and linting error. CI should be passed.

**Comment by plroebuck** [comment]
@outsideris, @boneskull,

issue #3569 may be implicating this PR. Can you confirm?
If so, we may need to back this out...

**Comment by plroebuck** [comment]
@boneskull, this needs to get unmerged until the order is corrected. See #3569.

**Comment by boneskull** [comment]
I'll reverse this.

**Comment by papb** [comment]
@outsideris Hello, I am facing this issue... You seem to have an idea on how to solve it:

> We should check `afterEach` hooks to support conditional fail and report it in a correct order.

How can I help?
